File: jscomp/others/bs_Queue.ml
(**************************************************************************)
(*                                                                        *)
(*                                 OCaml                                  *)
(*                                                                        *)
(*         Francois Pottier, projet Cristal, INRIA Rocquencourt           *)
(*                  Jeremie Dimino, Jane Street Europe                    *)
(*                                                                        *)
(*   Copyright 2002 Institut National de Recherche en Informatique et     *)
(*     en Automatique.                                                    *)
(*                                                                        *)
(*   All rights reserved.  This file is distributed under the terms of    *)
(*   the GNU Lesser General Public License version 2.1, with the          *)
(*   special exception on linking described in the file LICENSE.          *)
(*                                                                        *)
(**************************************************************************)
(* Adapted siginifcantly by BuckleScript Authors                          *)
type 'a node = { 
  content: 'a; 
  mutable next: 'a cell 
}
and 'a cell = 'a node Js.null
and 'a t = {
  mutable length: int;
  mutable first: 'a cell;
  mutable last: 'a cell
} [@@bs.deriving abstract]

let null  = Js.null 
let return = Js.Null.return 

let create () = 
  t 
    ~length: 0
    ~first:null
    ~last:null


let clear q =
  lengthSet q  0;
  firstSet q  null;
  lastSet q  null

let push q x =
  let cell = return @@ node 
      ~content:x
      ~next:null
  in
  match Js.nullToOption (last q )with
  | None ->
    lengthSet  q 1;
    firstSet q cell;
    lastSet q cell
  | Some last ->
    lengthSet q (length q + 1);
    nextSet last  cell;
    lastSet q  cell


let peekOpt q =
  match Js.nullToOption (first q ) with
  | None -> None
  | Some v -> Some (content v)

let peekNull q =
  match Js.nullToOption (first q ) with
  | None -> null
  | Some v -> return (content v)


let peekAssert q =
  match Js.nullToOption (first q ) with
  | None -> [%assert "Bs.Queue.Empty"]
  | Some v -> content v

let popOpt q =
  match Js.nullToOption (first q ) with
  | None -> None
  | Some x  ->
    let next = next x in 
    if Js.Null.test next then 
      begin (* only one element*)
        clear q;
        Some (content x)
      end
    else begin 
      lengthSet q (length q - 1);
      firstSet q next;
      Some(content x) 
    end

let popAssert q =
  match Js.nullToOption (first q ) with
  | None -> [%assert "Bs.Queue.Empty"]
  | Some x  ->
    let next = next x in 
    if Js.Null.test next then 
      begin (* only one element*)
        clear q;
        content x
      end
    else begin 
      lengthSet q (length q - 1);
      firstSet q next;
      content x 
    end    

let popNull q =
  match Js.nullToOption (first q ) with
  | None -> null
  | Some x  ->
    let next = next x in 
    if Js.Null.test next then 
      begin (* only one element*)
        clear q;
        return (content x)
      end
    else begin 
      lengthSet q (length q - 1);
      firstSet q next;
      return (content x) 
    end

let rec copyAux qRes prev cell =
  match Js.nullToOption cell with
  | None -> lastSet qRes  prev; qRes
  | Some x  ->
    (* Cons { content; next } *)
    let content = content x in 
    let res = return @@ node ~content ~next:null in
    begin match Js.nullToOption prev with
      | None -> firstSet qRes res
      | Some p -> nextSet p  res
    end;
    copyAux qRes res (next x)

let copy q =

  copyAux (t  ~length:(length q) ~first:null ~last:null)  null (first q)

let isEmpty q =
  length q = 0

let length q =
  length q

let rec iterAux f cell =
  match Js.nullToOption cell with
  | None -> ()
  | Some x  ->
    f (content x) [@bs];
    iterAux f (next x)

let iter q f =
  iterAux f (first q)

let rec foldAux f accu cell =
  match Js.nullToOption cell with
  | None -> accu
  | Some x  ->
    let accu = f accu (content x) [@bs] in
    foldAux f accu (next x)

let fold q  accu f =
  foldAux f accu (first q)

let transfer q1 q2 =
  if length q1 > 0 then
    match Js.nullToOption (last q2) with
    | None ->
      lengthSet q2 (length q1);
      firstSet q2 (first q1);
      lastSet q2 (last q1);
      clear q1
    | Some l ->
      lengthSet q2  (length q2 + length q1);
      nextSet l (first q1);
      lastSet q2 (last  q1);
      clear q1

let rec fillAux i arr cell =       
  match Js.nullToOption cell with 
  | None -> ()
  | Some x ->
    Bs_Array.unsafe_set arr i (content x);
    fillAux (i + 1) arr (next x) 

let toArray x =         
  let v = Bs_Array.makeUninitializedUnsafe (length x) in 
  fillAux 0 v (first x);
  v

File: lib/js/bs_Queue.js
'use strict';


var $$null = null;

function create() {
  return {
          length: 0,
          first: $$null,
          last: $$null
        };
}

function clear(q) {
  q.length = 0;
  q.first = $$null;
  q.last = $$null;
  return /* () */0;
}

function push(x, q) {
  var cell = {
    content: x,
    next: $$null
  };
  var match = q.last;
  if (match !== null) {
    q.length = q.length + 1 | 0;
    match.next = cell;
    q.last = cell;
    return /* () */0;
  } else {
    q.length = 1;
    q.first = cell;
    q.last = cell;
    return /* () */0;
  }
}

function peekOpt(q) {
  var match = q.first;
  if (match !== null) {
    return /* Some */[match.content];
  } else {
    return /* None */0;
  }
}

function peekNull(q) {
  var match = q.first;
  if (match !== null) {
    return match.content;
  } else {
    return $$null;
  }
}

function peekAssert(q) {
  var match = q.first;
  if (match !== null) {
    return match.content;
  } else {
    throw new Error("Bs.Queue.Empty");
  }
}

function popOpt(q) {
  var match = q.first;
  if (match !== null) {
    var next = match.next;
    if (next === null) {
      clear(q);
      return /* Some */[match.content];
    } else {
      q.length = q.length - 1 | 0;
      q.first = next;
      return /* Some */[match.content];
    }
  } else {
    return /* None */0;
  }
}

function popAssert(q) {
  var match = q.first;
  if (match !== null) {
    var next = match.next;
    if (next === null) {
      clear(q);
      return match.content;
    } else {
      q.length = q.length - 1 | 0;
      q.first = next;
      return match.content;
    }
  } else {
    throw new Error("Bs.Queue.Empty");
  }
}

function popNull(q) {
  var match = q.first;
  if (match !== null) {
    var next = match.next;
    if (next === null) {
      clear(q);
      return match.content;
    } else {
      q.length = q.length - 1 | 0;
      q.first = next;
      return match.content;
    }
  } else {
    return $$null;
  }
}

function copy(q) {
  var qRes = {
    length: q.length,
    first: $$null,
    last: $$null
  };
  var _prev = $$null;
  var _cell = q.first;
  while(true) {
    var cell = _cell;
    var prev = _prev;
    if (cell !== null) {
      var content = cell.content;
      var res = {
        content: content,
        next: $$null
      };
      if (prev !== null) {
        prev.next = res;
      } else {
        qRes.first = res;
      }
      _cell = cell.next;
      _prev = res;
      continue ;
      
    } else {
      qRes.last = prev;
      return qRes;
    }
  };
}

function isEmpty(q) {
  return +(q.length === 0);
}

function length(q) {
  return q.length;
}

function iter(f, q) {
  var f$1 = f;
  var _cell = q.first;
  while(true) {
    var cell = _cell;
    if (cell !== null) {
      f$1(cell.content);
      _cell = cell.next;
      continue ;
      
    } else {
      return /* () */0;
    }
  };
}

function fold(f, accu, q) {
  var f$1 = f;
  var _accu = accu;
  var _cell = q.first;
  while(true) {
    var cell = _cell;
    var accu$1 = _accu;
    if (cell !== null) {
      var accu$2 = f$1(accu$1, cell.content);
      _cell = cell.next;
      _accu = accu$2;
      continue ;
      
    } else {
      return accu$1;
    }
  };
}

function transfer(q1, q2) {
  if (q1.length > 0) {
    var match = q2.last;
    if (match !== null) {
      q2.length = q2.length + q1.length | 0;
      match.next = q1.first;
      q2.last = q1.last;
      return clear(q1);
    } else {
      q2.length = q1.length;
      q2.first = q1.first;
      q2.last = q1.last;
      return clear(q1);
    }
  } else {
    return 0;
  }
}

function fillAux(_i, arr, _cell) {
  while(true) {
    var cell = _cell;
    var i = _i;
    if (cell !== null) {
      arr[i] = cell.content;
      _cell = cell.next;
      _i = i + 1 | 0;
      continue ;
      
    } else {
      return /* () */0;
    }
  };
}

function toArray(x) {
  var v = new Array(x.length);
  fillAux(0, v, x.first);
  return v;
}

exports.clear      = clear;
exports.create     = create;
exports.push       = push;
exports.peekOpt    = peekOpt;
exports.peekNull   = peekNull;
exports.peekAssert = peekAssert;
exports.popOpt     = popOpt;
exports.popNull    = popNull;
exports.popAssert  = popAssert;
exports.copy       = copy;
exports.isEmpty    = isEmpty;
exports.length     = length;
exports.iter       = iter;
exports.fold       = fold;
exports.transfer   = transfer;
exports.toArray    = toArray;
/* null Not a pure module */
